Cohort as a query term

f-marino opened the pull request:
Defined API to allow the use of cohort as query term.
To do that, the client is now allowed to pass patient set IDs as panel items.

Work to be done on the backend:
- minor adjustments in the i2b2 wrapper and on the cli to manage the newly defined `cohortItems` (and `ConceptItems`) fields of the `/query` API.
- Estimated time ~1 week.

Work to be done on the frontend:
- making the cohorts (from the side-panel module) drag-and-droppable on the explore-module. 
- update the current calls to the `/query` endpoint.
- Estimated time (for Jimmy) 2-3 weeks.

f-marino committed:
defined APIs to implement the cohort-as-a-query term feature

mickmis commented on connector/swagger/medco-connector.yml:
```diff
@@ -849,13 +849,14 @@ definitions:
         type: "boolean"
         description: "exclude the i2b2 panel"
-      items:
+      conceptItems:
```
Here the term *item* was set on purpose in order to match the i2b2 terminology. An i2b2 query is composed of *panels* in which there are *items*, which can be *concepts*, *modifiers*, soon *cohorts*, or also other stuff available in the i2b2 API. I believe it still makes sense to keep this consistency, notably because in any case (well, as long as we keep the i2b2 software) the query format of MedCo is constrained by the query format of i2b2.

So I would suggest to keep the terminology aligned with i2b2: `conceptItems` -> `items`

f-marino replied:
I agree that ideally here we should have an array of `item`, with `item` being either a concept or a patient set id. Unfortunately with OpenAPI v2.0 it's not possible to do that, so I provided separate arrays of concept items and patient set id items. I know that under the hood everything will end up in the `<item_key>` field of i2b2, but this way is easier to perform consistency check at the API level (a concept item must contain a path and other optional fields, a patient set id item must contain an integer and that's it)

mickmis replied:
OK I see what you mean, I still don't agree though as I don't find that consistent, but it's your call now.

mickmis commented on connector/swagger/medco-connector.yml:
```diff
@@ -914,6 +915,11 @@ definitions:
                   pattern: '^((\/[^\/]+)+\/)$'
             encrypted:
               type: "boolean"
+      patientSetIDItems:
```
Linked to the other comment, since a patient set is just a type of item for i2b2, I would say to change this way of specifying the patient set ID, and put it along with the other query terms.
Also right now, it does not allow mixing together patient set IDs with other query terms (concepts or modifiers), which I believe is required (and enabled by i2b2).

I would suggest to:
- remove the array and make a simple field
- alternatively:
  1. put this as a field of `items` / `conceptItems` alongside `queryTerm`, `modifier`, etc.
  2. remove this field altogether and use the field `queryTerm` for specifying the patient set ID, and add a specific operator (e.g. `PATIENT_SET`) and/or add a `type` (e.g. `PATIENT_SET`)

f-marino replied:
> Linked to the other comment, since a patient set is just a type of item for i2b2, I would say to change this way of specifying the patient set ID, and put it along with the other query terms.

Linked to my previous comment, this way we would lose the consistency check at the API level.

> Also right now, it does not allow mixing together patient set IDs with other query terms (concepts or modifiers), which I believe is required (and enabled by i2b2).

Actually it's possible, `conceptItems` and `patientSetIDItems `are both under `panel`

f-marino committed:
implemented server and client managing patientSetID items

f-marino committed:
added tests for patient set id as query term

f-marino committed:
refactored cli tests

f-marino commented:
Changes summary:
- changed `panel` definition to distinguish between `ConceptItems` and `PatientSetIDItems`.
- changes in the i2b2 wrapper to manage the two kind of items.
- changes in the CLI to manage the two kind of items. WARNING: the CLI assumes that the patient set IDs belonging to the same cohort are the same for all nodes. This is not true in general, but it's always true in our test settings. Given the complexity that would have entailed to manage different patient set IDs at the CLI side I preferred to adopt this simplified solution, but let me know if you think otherwise.
- added tests for the new feature ([here](https://github.com/ldsec/medco/blob/688990c8789c7f23cc5d65706c5a404ae0906ea4/test/test_cli_preloading/query.sh#L28) the link for commodity).
- split of the cli tests in multiple files for better manageability.

mickmis commented on connector/client/parse_query.go:
```diff
@@ -86,74 +86,90 @@ func ParseQueryString(queryString string) (panels []*models.Panel, err error) {
 }
 
-// ParseQueryItem parses a string into an item
+// ParseQueryItem parses a string into an item.
```
The [MedCo Documentation](https://ldsec.gitbook.io/medco-documentation/system-administrators/cli) should be updated accordingly, either when merged, or a v2.0.2 branch can already be created.

mickmis reviewed (approved):
> the CLI assumes that the patient set IDs belonging to the same cohort are the same for all nodes. This is not true in general, but it's always true in our test settings. Given the complexity that would have entailed to manage different patient set IDs at the CLI side I preferred to adopt this simplified solution, but let me know if you think otherwise.

If implemented that way, this requirement should be explicitly specified (e.g. doc, readme or something else) so that it's not left as something unspecified that may break because of a future change.

Also, an alternative to that could be that the API changes so that instead of accepting the patient ID (different for each node), it accepts the cohort name (unique for all nodes), saving any client to do the mapping.

f-marino committed:
fixed inconsistence in API field naming

f-marino committed:
changed API to accept cohort names instead of patient set IDs

f-marino commented:
> Also, an alternative to that could be that the API changes so that instead of accepting the patient ID (different for each node), it accepts the cohort name (unique for all nodes), saving any client to do the mapping.

I like that, I updated the API, the code and the PR description accordingly. Merge at will.

f-marino committed:
fixed naming inconsistence

jimmydjabali commented:
@f-marino As the Front-end PR's has been merged (https://github.com/ldsec/glowing-bear-medco/pull/217), if yours is ready, you can also merge it 🙂

f-marino committed:
Merge branch 'dev' into cohort-as-query-term
